text/0000-const-generics.md

```diff
@@ -0,0 +1,287 @@
+well all idents - e.g. `3`, `"Hello, world"`, `foo_bar`.
+
+## Declaring a const parameter
+
+In any sequence of type parameter declarations (such as in the definition of a
+type or on the `impl` header of an impl block) const parameters can also be
+declared. Const parameters always come after type parameters, and their
+declarations take the form `const $ident: $ty`, as in:
+
+```rust
+struct RectangularArray<T, const WIDTH: usize, const HEIGHT: usize> {
+    array: [[T; WIDTH]; HEIGHT],
+}
+```
+
+The idents declared are the names used for these const parameters
+(interchangeably called "const variables" in this RFC text), and all values
+must be of the type ascribed to it. Which types can be ascribed to const
+parameters is restricted later in this RFC.
+
+The const parameter is in scope for the entire body of the item (type, impl,
+function, method, etc) in which it is declared.
+
+## Applying a const as a parameter
+
+Any const expression of the type ascribed to a const parameter can be applied
+as that parameter. When applying an expression as const parameter (except for
+arrays), which is not an identity expression, the expression must be contained
+within a block. This syntactic restriction is necessary to avoid requiring
+infinite lookahead when parsing an expression inside of a type.
+
+```rust
+const X: usize = 7;
+
+let x: RectangularArray<i32, 2, 4>;
+let y: RectangularArray<i32, X, {2 * 2}>;
+```
+
+### Arrays
+Arrays have a special construction syntax: `[T; CONST]`. In array syntax,
+braces are not needed around any const expressions; `[i32; N * 2]` is a
+syntactically valid type.
+
+## When a const variable can be used
+
+A const variable can be used as a const in any of these contexts:
+
+1. As an applied const to any type which forms a part of the signature of
+the item in question: `fn foo<const N: usize>(arr: [i32; N])`.
+2. As part of a const expression used to define an associated const, or as a
+parameter to an associated type.
+3. As a value in any runtime expression in the body of any functions in the
+item.
+4. As a parameter to any type used in the body of any functions in the item,
+as in `let x: [i32; N]` or `<[i32; N] as Foo>::bar()`.
+5. As a part of the type of any fields in the item (as in
+`struct Foo<const N: usize>([i32; N]);`).
+
+In general, a const variable can be used where a const can. There is one
+significant exception: const variables cannot be used in the construction of
+consts, statics, functions, or types inside a function body. That is, these
+are invalid:
+
+```rust
+fn foo<const X: usize>() {
+    const Y: usize = X * 2;
+    static Z: (usize, usize)= (X, X);
+
+    struct Foo([i32; X]);
+}
+```
+
+This restriction can be analogized to the restriction on using type variables
+in types constructed in the body of functions - all of these declarations,
+though private to this item, must be independent of it, and do not have any
+of its parameters in scope.
```

Comment:
Yeah I was only thinking of consts. Since they're basically just named temporaries, it seems totally fine (no weird codegen implications like statics).